src/core/shape_inference/include/topk_shape_inference.hpp

```diff
@@ -13,12 +13,44 @@ namespace ov {
 namespace op {
 namespace v1 {
 
-template <typename T>
-void shape_infer(const TopK* op,
-                 const std::vector<T>& input_shapes,
-                 std::vector<T>& output_shapes,
-                 const std::map<size_t, std::shared_ptr<ngraph::runtime::HostTensor>>& constant_data = {}) {
-    NODE_VALIDATION_CHECK(op, (input_shapes.size() == 2 && output_shapes.size() == 2));
+// Helper to get correct K from tensor as shape.
+template <class T>
+struct GetK {
+    const TopK* m_op;
+
+    GetK(const TopK* op) : m_op{op} {}
+
+    template <class K>
+    T operator()(const K k) const {
+        NODE_VALIDATION_CHECK(m_op,
+                              cmp::ge(k, 0) && cmp::le(k, std::numeric_limits<T>::max()),
+                              "The value of 'K' must be more or equal zero.",
+                              " (got ",
+                              k,
+                              ").");
+        return sh_infer::tr::Cast<T>()(k);
```

Comment:
Since the `Cast` structure doesn't perform any validation of the types or the value of `k` I think using `static_cast<T>` could be used here directly.

Reply:
Done.